### PR TITLE
Feature/metagenomics exchange

### PR DESCRIPTION
**Pull request by KateSakharova**
Scripts to populate and track Metagenomics Exchange https://www.ebi.ac.uk/ena/registry/metagenome/api/
ME is based on EBI Search model.
System for ME works a bit differently. For update process we do not need to remove and then add a record to "XML" again. That is why I do not use `to_delete()` function.
I've created `get_suppressed()` to filter records that need to be removed from ME API.

### CODE REVIEW

File: emgapi/metagenomics_exchange.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+            "Authorization": self.__token,
+        }
+        response = requests.patch(
+            f"{self.base_url}/{endpoint}", json=data, headers=headers
+        )
+        response.raise_for_status()
+        return response
+
+    def add_analysis(self, mgya: str, run_accession: str, public: bool):
+        data = {
```
[the remark this answers is not in the document]
Comment (mberacochea) [reply emgapi/metagenomics_exchange.py:77]: I would keep one copy of this, probably here

**Review by mberacochea — CHANGES_REQUESTED** [review]
Great stuff Kate, I left some comments :)

File: emgapi/management/commands/populate_metagenomics_exchange.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                            # Just to be safe, update the MGX accession
+                            ajob.mgx_accession = registry_id
+                            ajob.last_mgx_indexed = timezone.now()
+                            jobs_to_update.append(ajob)
+                        else:
+                            logging.error(f"Analysis {ajob} update failed")
+                    else:
+                        logging.debug(f"No edit for {ajob}, metadata is correct")
+
+            AnalysisJob.objects.bulk_update(
```
[the remark this answers is not in the document]
Comment (KateSakharova) [reply emgapi/management/commands/populate_metagenomics_exchange.py:167]: It is already for every page

[the remark this answers is not in the document]
Comment (mberacochea) [reply emgapi/management/commands/populate_metagenomics_exchange.py:154]: ahh... sorry, ignore this then :facepalm:

File: emgapi/models.py [hunk 3]
```diff
@@ -1634,6 +1687,27 @@ def __str__(self):
...
+        db_column='LAST_POPULATED_ME',
+        null=True,
+        blank=True,
+        help_text="Date at which this model was last appeared in Metagenomics Exchange"
+    )
+
+    objects_for_population = MetagenomicsExchangeQueryset.as_manager()
+
+    class Meta:
+        abstract = True
```
Comment (SandyRogers) [inline emgapi/models.py:1708]: What is this abstract model for? Isn't `last_populated_me` now done by `last_mgx_indexed` in the `MetagenomicsExchangeIndexedModel`? Maybe I misunderstood.

File: emgcli/settings.py [hunk 4]
```diff
@@ -682,3 +682,16 @@ def create_secret_key(var_dir):
...
+METAGENOMICS_EXCHANGE_API = ""
+METAGENOMICS_EXCHANGE_API_TOKEN = ""
+try:
+    METAGENOMICS_EXCHANGE_API = EMG_CONF['emg']['me_api']
+    if EMG_CONF['emg'].get('me_api_token'):
+        METAGENOMICS_EXCHANGE_API_TOKEN = EMG_CONF['emg']['me_api_token']
+    else:
+        METAGENOMICS_EXCHANGE_API_TOKEN = os.getenv('METAGENOMICS_EXCHANGE_API_TOKEN')
+except KeyError:
+    warnings.warn("The metagenomics exchange API and Token are not configured properly")
```
Comment (SandyRogers) [inline emgcli/settings.py:697]:
```suggestion
try:
    METAGENOMICS_EXCHANGE_API = EMG_CONF['emg']['me_api']
    METAGENOMICS_EXCHANGE_API_TOKEN = os.getenv('METAGENOMICS_EXCHANGE_API_TOKEN')
    if not METAGENOMICS_EXCHANGE_API_TOKEN:
        METAGENOMICS_EXCHANGE_API_TOKEN = EMG_CONF['emg']['me_api_token']
except KeyError:
    warnings.warn("The metagenomics exchange API and Token are not configured properly")
```

Just to make sure that a missing token (not in env or in config) triggers the warning.

File: emgapi/migrations/0017_auto_20240129_1401.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        ),
+        migrations.RemoveField(
+            model_name='study',
+            name='last_indexed',
+        ),
+        migrations.AddField(
+            model_name='analysisjob',
+            name='last_ebi_search_indexed',
+            field=models.DateTimeField(blank=True, db_column='LAST_EBI_SEARCH_INDEXED', help_text='Date at which this model was last included in an EBI Search initial/incremental index.', null=True),
+        ),
```
Comment (SandyRogers) [inline emgapi/migrations/0017_auto_20240129_1401.py:25]: Can we have the migration to a `RenameField` operation instead please, on the Study and AnalysisJob `last_indexed` fields? That way, we won't lose track of the existing EBI Indexation state.

File: emgapi/metagenomics_exchange.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+from django.conf import settings
+
+
+class MetagenomicsExchangeAPI:
+
+    """Metagenomics Exchange API Client"""
+
+    def __init__(self, base_url=None):
+        self.base_url = base_url if base_url else settings.METAGENOMICS_EXCHANGE_API
```
Comment (SandyRogers) [inline emgapi/metagenomics_exchange.py:28]:
```suggestion
        self.base_url = base_url or settings.METAGENOMICS_EXCHANGE_API
```

File: emgapi/management/commands/populate_metagenomics_exchange.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright 2017-2022 EMBL - European Bioinformatics Institute
```
Comment (SandyRogers) [inline emgapi/management/commands/populate_metagenomics_exchange.py:4]:
```suggestion
# Copyright 2017-2024 EMBL - European Bioinformatics Institute
```

File: emgapi/metagenomics_exchange.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright 2018-2023 EMBL - European Bioinformatics Institute
```
Comment (SandyRogers) [inline emgapi/metagenomics_exchange.py:4]:
```suggestion
# Copyright 2018-2024 EMBL - European Bioinformatics Institute
```

File: emgapi/management/commands/populate_metagenomics_exchange.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        self.process_to_index_and_update_records(analyses_to_index_and_update)
+        self.process_to_delete_records(analyses_to_delete)
+
+        logging.info("Done")
+
+    def process_to_index_and_update_records(self, analyses_to_index_and_update):
+        logging.info(f"Indexing {len(analyses_to_index_and_update)} new analyses")
+
+        for page in Paginator(analyses_to_index_and_update, 100):
```
Comment (SandyRogers) [inline emgapi/management/commands/populate_metagenomics_exchange.py:99]: This `100` is probably worth making a parameter – we had some trouble with EBI dump where reasonable sounding page sizes still caused query problem on the DB. If parametrized we can at least easily try different values.

File: emgapi/management/commands/populate_metagenomics_exchange.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+                                f"Dry-mode run: no patch to real ME for {ajob}"
+                            )
+                            continue
+                        if self.mgx_api.patch_analysis(
+                                registry_id=registry_id, data=metadata
+                        ):
+                            logging.info(f"Analysis {ajob} updated successfully")
+                            # Just to be safe, update the MGX accession
+                            ajob.mgx_accession = registry_id
+                            ajob.last_mgx_indexed = timezone.now() + timedelta(minutes=1)
```
Comment (SandyRogers) [inline emgapi/management/commands/populate_metagenomics_exchange.py:147]: One thing to consider here, is that if we offset individual jobs indexation by only 1 minute individually, by the time this loop exists and the bulk_update on L154 happens – that +minute may be in the past. This means that the auto last_update will now be after the last indexation (if Django decided to update the last_update for bulk updates... which I think it currently doesn't, but might in future). Perhaps worth a quick test / check the Django docs to be sure this won't be a problem.

**Review by SandyRogers — CHANGES_REQUESTED** [review]
Thank you @KateSakharova , great stuff! Couple of small suggestions and queries inline.

File: emgapi/migrations/0017_auto_20240129_1401.py (hunk 5 above)
Comment (mberacochea) [reply emgapi/migrations/0017_auto_20240129_1401.py:25]: Yep, the PR is "outdated" now.

File: emgapi/models.py (hunk 3 above)
Comment (KateSakharova) [reply emgapi/models.py:1708]: good catch *facepalm*, it was left after ebi_search merge by mistake, removed

**Review by SandyRogers — APPROVED** [review]
Thank you @KateSakharova !

**Review by mberacochea — APPROVED** [review]
Thanks Kate!

**Pull request merged**